sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -310,3 +310,97 @@ object PhysicalWindow {
     case _ => None
   }
 }
+
+/**
+ * Extract partition push down condition from ExpressionSet
+ * Since origin judge condition is
+ *    {
+ *       !expression.references.isEmpty &&
+ *             expression.references.subsetOf(partitionKeyIds)
+ *    }
+ *
+ *  This can only push down simple condition expression.
+ *  Such as table:
+ *   CREATE TABLE DEFAULT.PARTITION_TABLE(
+ *   A STRING,
+ *   B STRING)
+ *   PARTITIONED BY(DT STRING)
+ *
+ * With SQL:
+ *   SELECT A, B
+ *   FROM DEFAULT.PARTITION_TABLE
+ *   WHERE DT = 20190601 OR (DT = 20190602 AND C = "TEST")
+ *
+ * Where condition "DT = 20190601 OR (DT = 20190602 AND C = "TEST")"
```

Comment:
Let's be more clear about the approach here. We try to weaken a predicate so that it only refers to partition columns, by removing some branches in AND. Let's also mention the corner case when there is no partition column attribute in the predicate(we should return Nil in this case?).

Reply:
@cloud-fan  
In this place it will return (DT = 20190601 OR DT = 20190602). 
But this whole condition will still return to filter.
What I want to do is purely to avoid  read unnecessary partitions. When this case we only read partition(dt=20190601 + dt=20190602), If we don't push down this, we will read all data.

In  condition " **(DT = 20190602 AND C = "TEST")** ", **DT = 20190602**  is **C = "TEST"**'s precondition. 

If the whole condition is **DT = 20190601 OR (DT = 20190602 OR C = "TEST")**. We should return null, since **DT = 20190602**  is not  **C = "TEST"**'s  constraint.

Reply:
@cloud-fan 
In my code. coming predicate Set[Expression] has a potential **AND** logical.
For one Expression, it will be   restricted by other same level Expression.
and  :

- if it is a combine of **AND** each side can be a constraint to others, so it one side is tenable, it can return a tenable condition.
- if it is a combine of **OR**, if one side is out of control(such as have no condition about partition cols) this whole **OR** Expression should return NONE.  Only  when both side of **OR** 's child is reasonable, it can return a  tenable combine of **OR**. 
- if it 's a multilayer nested Expression combined by BinaryOperator.  It will visit the lowest level, if it found one level's **OR** Expression is untenable, it will break this Expression totally and return null.